lib/si-pool-noodle/src/pool_noodle.rs

```diff
@@ -76,233 +107,185 @@ where
+            max_concurrency: config.max_concurrency,
+            ready_queue: ArrayQueue::new(config.pool_size as usize),
+            retry_limit: config.retry_limit,
+            shutdown_token: config.shutdown_token,
+            spec: config.spec,
+            work_queue,
         }));
-        for n in 1..=pool_size {
+        for id in 1..=config.pool_size {
             let me = Arc::clone(&pool.0);
-            Self::push_to_clean(me, n);
+            Self::push_clean_task_to_work_queue(me, id);
         }
         pool
     }
 
-    /// Gets the current pool stats from the inner struct
-    /// Only used for tests
-    pub async fn stats(&self) -> PoolNoodleStats {
-        self.0.stats().await
-    }
-
-    /// Start PoolNoodle. It will spin up various threads to handle Cyclone Instance lifecycles.
-    #[allow(clippy::let_underscore_future)] // These needs to just run in the background forever.
-    pub fn start(&mut self, check_health: bool) -> Result<()> {
-        if check_health {
+    /// do the thing
+    pub fn run(&mut self) -> Result<()> {
+        if self.inner().check_health {
             if let Some(err) = futures::executor::block_on(self.check_health()).err() {
                 return Err(err);
             }
         }
+        let me = self.inner();
+        let semaphore = Arc::new(Semaphore::new(self.inner().max_concurrency as usize));
 
-        let me = Arc::clone(&self.0);
-
-        // TODO(nick,fletcher,scott): shift towards our tokio select, task tracker and cancellation
-        // token pattern.
-        for _ in 0..INTERNAL_WORKERS_COUNT {
-            let _ = tokio::spawn(Self::handle_prepare(me.clone()));
-
-            let _ = tokio::spawn(Self::handle_clean(me.clone()));
-
-            let _ = tokio::spawn(Self::handle_drop(me.clone()));
-        }
-        Ok(())
-    }
-
-    async fn handle_prepare(me: Arc<PoolNoodleInner<I, S>>) {
-        debug!("PoolNoodle: starting prepare handler...");
-
-        let inner = async {
+        tokio::spawn(async move {
             loop {
-                // let's make more instances!
-                if let Some(id) = Self::pop_from_unprepared(me.clone()) {
-                    debug!("PoolNoodle: readying instance");
-                    match PoolNoodleInner::prepare(id, &me.spec).await {
-                        Ok(_) => {
-                            debug!("PoolNoodle: instance readied: {}", id);
-                            match PoolNoodleInner::spawn(id, &me.spec).await {
-                                Ok(instance) => {
-                                    debug!("PoolNoodle: instance started: {}", id);
-                                    Self::push_to_ready(me.clone(), instance);
+                tokio::select! {
+                    _ = me.shutdown_token.cancelled() => {
+                        debug!("main loop received cancellation");
+                        break
+                    }
+                    Ok(permit) = semaphore.acquire() => {
```

Comment:
We discussed potential failure modes here and @sprutton1 is going to confirm if this works as intended